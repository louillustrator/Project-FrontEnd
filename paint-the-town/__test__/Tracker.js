import React from "react";
import Tracker from "../components/Tracker";
import { shallow } from 'enzyme';

import renderer from "react-test-renderer";

test('renders without crashing', () => {
    const rendered = renderer.create(<Tracker />).toJSON();
    expect(rendered).toBeTruthy();
});

test("renders correctly", () => {
    const tree = renderer.create(<Tracker />).toJSON();
    expect(tree).toMatchSnapshot();
});
