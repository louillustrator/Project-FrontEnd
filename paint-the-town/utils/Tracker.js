import haversine from 'haversine';
import { Location, Permissions, MapView } from "expo";

export const _haversine = (coordArray) => {
    if (coordArray.length > 1) {
        let metometer = 0;
        for (let i = 0; i < coordArray.length - 1; i++) {
            metometer += haversine(coordArray[i], coordArray[i + 1], { unit: 'meter' })
            //other measurements: 'km'(default), 'mile', 'nmi'
        }
        return metometer
    }
}

export const _watchPosition = function(val) {
    if(val === 0 && this.state.watching === true) {
        this.state.subscription.remove()
    } else if(val === 1 && this.state.watching === false)
    Permissions.askAsync(Permissions.LOCATION)
      .then(res => {
        if (res.status === "granted") {
          returnedPromise = Location.watchPositionAsync(
            {
              enableHighAccuracy: true,
              distanceInterval: 2,
              timeInterval: 2000
            },
            newLocation => {
              if (newLocation.timestamp) {
                let object = {
                  latitude: newLocation.coords.latitude,
                  longitude: newLocation.coords.longitude,
                  timestamp: newLocation.timestamp
                };
                let route = [...this.state.route]
                let currentObj = {...route[route.length -1]}
                let latLng = [...currentObj.latLng, object]
                currentObj.latLng = latLng
                route[route.length-1] = currentObj
                let subscription = returnedPromise._55
                this.setState({
                  route,
                  subscription
                });
              } else {
                console.log("ignored newLocation");
              }
            }
          );
        }
      })
      .catch(err => console.log("Error in componentDidMount: ", err.message));
  };

  export const _getLocationAsync = async function() {
    let { status } = await Permissions.askAsync(Permissions.LOCATION);
    if (status !== "granted") {
      this.setState({
        errorMessage: "Permission to access location was denied"
      });
    }
    let location = await Location.getCurrentPositionAsync({
      enableHighAccuracy: true
    });

    return location;
  };