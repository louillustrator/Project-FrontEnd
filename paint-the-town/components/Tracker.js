import React, { Component } from "react";
import { View, Switch, Text } from "react-native";
import PropTypes from 'prop-types';
import { MapView } from "expo";
import ActionButt from "./ActionButt";
import * as tracker from "../utils/Tracker";
import Polylines from "./Polylines";
import * as api from "../utils/api";
import { takeSnapshotAsync } from "expo";
import { trackerStyles } from "../styles";
import WidthPicker from "./WidthPicker";

class Tracker extends Component {
  state = {
    region: {
      latitude: 55.3781,
      longitude: 3.436,
      latitudeDelta: 0.01,
      longitudeDelta: 0.01
    },
    errorMessage: null,
    subscrition: {},
    route: [{ latLng: [], colour: "#3600ff", width: 1 }],
    width: 1,
    status: null,
    watching: false,
    colour: "#3600ff",
    blueDot: true,
    toggle: false,
    showSlider: false,
    loadedMap: false
  };

  componentDidMount = async () => {
    let location = await this._getLocationAsync();
    this.setState({
      status: "granted",
      region: {
        ...this.state.region,
        latitude: location.coords.latitude,
        longitude: location.coords.longitude
      }
    });
    let journey = this.props.navigation.getParam("journey");
    if (journey)
      this.setState({
        route: journey.route,
        loadedMap: true
      });
  };

  _getLocationAsync = tracker._getLocationAsync.bind(this);
  _watchPosition = tracker._watchPosition.bind(this);

  render() {
    const {
      container,
      navBar,
      navBarText,
      text,
      map,
      butt,
      toggleButt
    } = trackerStyles;
    const {
      toggle,
      region,
      blueDot,
      route,
      colour,
      width,
      showSlider
    } = this.state;

    return (
      <View style={container}>
        <View style={navBar}>
          <Text style={navBarText}>My Journey</Text>
        </View>

        <View style={toggleButt}>
          <Switch onValueChange={this.toggle} value={toggle} />
          <Text style={text}>Satellite View</Text>
        </View>

        <MapView
          collapsable={false}
          style={map}
          ref="routeMap"
          region={region}
          showsUserLocation={blueDot}
          showsMyLocationButton={true}
          showsCompass={true}
          followUserLocation={true}
          zoomEnabled={true}
          onRegionChangeComplete={this.setZoom}
          rotateEnabled={true}
          mapType={toggle ? "satellite" : "standard"}
        >
          <Polylines route={route} />
        </MapView>
        {showSlider && (
          <WidthPicker
            setShowSlider={this.setShowSlider}
            setWidth={this.setWidth}
            currentWidth={width}
          />
        )}
        <ActionButt
          style={butt}
          changeColour={this.changeColour}
          navigation={this.props.navigation}
          colour={colour}
          start={this.start}
          pause={this.pause}
          stop={this.stop}
          setShowSlider={this.setShowSlider}
        />
      </View>
    );
  }

  changeColour = async colour => {
    const { watching, width } = this.state;

    this.props.navigation.navigate("Tracker");
    if (watching) {
      let location = await this._getLocationAsync();
      let object = {
        latitude: location.coords.latitude,
        longitude: location.coords.longitude,
        timestamp: location.timestamp
      };
      let route = [...this.state.route];
      let currentObj = { ...route[route.length - 1] };
      let latLng = [...currentObj.latLng, object];
      currentObj.latLng = latLng;
      route[route.length - 1] = currentObj;
      let newObj = { latLng: [object], colour, width: this.state.width };
      route.push(newObj);
      this.setState({
        route,
        colour
      });
    } else {
      let route = [...this.state.route];
      let currentObj = { ...route[route.length - 1] };
      currentObj.colour = colour;
      route[route.length - 1] = currentObj;
      this.setState({
        route,
        colour
      });
    }
  };

  pause = async () => {
    this._watchPosition(0);
    let location = await this._getLocationAsync();
    let object = {
      latitude: location.coords.latitude,
      longitude: location.coords.longitude,
      timestamp: location.timestamp
    };
    let route = [...this.state.route];
    let currentObj = { ...route[route.length - 1] };
    let latLng = [...currentObj.latLng, object];
    currentObj.latLng = latLng;
    route[route.length - 1] = currentObj;
    let newObj = {
      latLng: [],
      colour: this.state.colour,
      width: this.state.width
    };
    route.push(newObj);
    this.setState({
      route,
      watching: false
    });
  };

  start = async () => {
    let location = await this._getLocationAsync();
    let object = {
      latitude: location.coords.latitude,
      longitude: location.coords.longitude,
      timestamp: location.timestamp
    };
    let route = [...this.state.route];
    if (!this.state.loadedMap) {
      let currentObj = { ...route[route.length - 1] };
      let latLng = [...currentObj.latLng, object];
      currentObj.latLng = latLng;
      route[route.length - 1] = currentObj;
      this._watchPosition(1);
      this.setState({
        route,
        watching: true
      });
    } else {
      let newObj = {
        latLng: [object],
        colour: this.state.colour,
        width: this.state.width
      };
      route = [...route, newObj];
      this._watchPosition(1);
      this.setState({
        route,
        loadedMap: false,
        watching: true
      });
    }
  };

  stop = async () => {
    const { currentUser } = this.props.screenProps;

    this._watchPosition(0);

    const newJourney = {
      user: currentUser,
      route: this.state.route
    };

    const newID = await api.storeJourney(newJourney);

    await this.setState({
      watching: false,
      blueDot: false
    });

    const newPic = await takeSnapshotAsync(this.refs.routeMap, {
      result: "base64",
      height: 1080,
      width: 1080,
      quality: 0,
      format: "jpeg"
    });
    api.storePic(newPic, newID, currentUser);
    this.props.navigation.navigate("Collection");
    this.setState({ blueDot: true });
  };

  setZoom = region => {
    this.setState({
      region
    });
  };

  toggle = () => this.setState({ toggle: !this.state.toggle });

  setShowSlider = val => {
    this.setState({
      showSlider: val
    });
  };

  setWidth = async width => {
    if (this.state.watching) {
      let location = await this._getLocationAsync();
      let object = {
        latitude: location.coords.latitude,
        longitude: location.coords.longitude,
        timestamp: location.timestamp
      };
      let route = [...this.state.route];
      let currentObj = { ...route[route.length - 1] };
      let latLng = [...currentObj.latLng, object];
      currentObj.latLng = latLng;
      route[route.length - 1] = currentObj;
      let newObj = { latLng: [object], colour: this.state.colour, width };
      route.push(newObj);
      this.setState({
        route,
        width
      });
    } else {
      let route = [...this.state.route];
      let currentObj = { ...route[route.length - 1] };
      currentObj.width = width;
      route[route.length - 1] = currentObj;
      this.setState({
        route,
        width
      });
    }
  };
}

Tracker.propTypes = {
  navigation: PropTypes.object.isRequired,
  screenProps: {
    currentUser: PropTypes.string.isRequired
  }
};

export default Tracker;
