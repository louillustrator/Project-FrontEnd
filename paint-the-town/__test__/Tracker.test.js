import React from "react";
import Tracker from "../components/Tracker";
import { View } from "react-native";
import { shallow } from 'enzyme';

import renderer from "react-test-renderer";


describe('Tracker', () => {
    let props;
    let state;
    beforeEach(() => {
        props = {
            screenProps: {
                currentUser: undefined,
                updateUser: undefined
            }
        };
        state = {
            region: {
                latitude: 55.3781,
                longitude: 3.436,
                latitudeDelta: 0.01,
                longitudeDelta: 0.01
            },
            errorMessage: null,
            subscrition: {},
            route: [{ latLng: [], colour: "#3600ff", width: 1 }],
            width: 1,
            status: null,
            watching: false,
            colour: "#3600ff",
            blueDot: true,
            toggle: false,
            showSlider: false
        };
        jest.mock('react-native-maps', () => 'MapView');
    });
    it('renders without crashing', () => {
        const rendered = renderer.create(<Tracker />).toJSON();
        expect(rendered).toBeTruthy();
    });
    it("renders correctly", () => {
        const tree = renderer.create(<Tracker />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("a view component is always rendered", () => {
        const wrapper = shallow(<Tracker />);
        expect(wrapper.find(View)).toHaveLength(3);
    });
    it('contains 2 children elements', () => {
        const wrapper = shallow(<Tracker />);
        expect(wrapper.find(View).children()).toHaveLength(7);
    });
});

