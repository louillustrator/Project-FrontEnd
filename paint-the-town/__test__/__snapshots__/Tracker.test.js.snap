// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tracker renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#63cdda",
        "height": 50,
        "position": "absolute",
        "width": 750,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "700",
          "textAlign": "center",
          "top": 15,
        }
      }
    >
      My Journey
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "flexWrap": "wrap",
        "position": "relative",
        "top": 65,
      }
    }
  >
    <RCTSwitch
      disabled={false}
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onValueChange={[Function]}
      style={
        Array [
          Object {
            "height": 31,
            "width": 51,
          },
          undefined,
        ]
      }
      value={false}
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 13,
          "fontStyle": "italic",
          "padding": 5,
        }
      }
    >
      Satellite View
    </Text>
  </View>
  <MapView
    collapsable={false}
    followUserLocation={true}
    mapType="standard"
    onRegionChangeComplete={[Function]}
    region={
      Object {
        "latitude": 55.3781,
        "latitudeDelta": 0.01,
        "longitude": 3.436,
        "longitudeDelta": 0.01,
      }
    }
    rotateEnabled={true}
    showsCompass={true}
    showsMyLocationButton={true}
    showsUserLocation={true}
    style={
      Object {
        "flex": 0.7,
        "height": 1334,
        "top": 90,
        "width": 750,
      }
    }
    zoomEnabled={true}
  >
    <View>
      <AIRMapPolyline
        coordinates={Array []}
        index={0}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeColors={Array []}
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={1}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={2}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={3}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={4}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={5}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={6}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={7}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={8}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={9}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={10}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={11}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={12}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={13}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={14}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={15}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={16}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={17}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={18}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={19}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
      <AIRMapPolyline
        index={20}
        lineCap="round"
        lineJoin="round"
        strokeColor="#000"
        strokeWidth={1}
      />
    </View>
  </MapView>
  <View
    pointerEvents="box-none"
    style={
      Object {
        "backgroundColor": "transparent",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "flexDirection": "column",
            "padding": 10,
          },
          Object {
            "alignItems": "center",
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
            },
            Object {
              "height": 63,
              "width": 63,
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "rgba(54, 0, 255, 1)",
                "borderRadius": 31.5,
                "height": 63,
                "justifyContent": "center",
                "shadowColor": "#444",
                "shadowOffset": Object {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.3,
                "shadowRadius": 1,
                "transform": Array [
                  Object {
                    "scale": 1,
                  },
                  Object {
                    "rotate": "0deg",
                  },
                ],
                "width": 63,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
